module/zfs/dmu_traverse.c

```diff
@@ -308,16 +314,51 @@ traverse_visitbp(traverse_data_t *td, const dnode_phys_t *dnp,
 
 		czb = kmem_alloc(sizeof (zbookmark_phys_t), KM_SLEEP);
 
+		/*
+		 * recursively visitbp() blocks below this.
+		 * Indirect (Non L0) Block of size 128k could contain, 1024
+		 * block pointers of 128 bytes. In case of full traverse OR
+		 * incremental traverse, where all blocks were modified, it
+		 * could traverse large number of blocks pointed by indirect.
+		 * Prefetching all blocks in go, could result into large number
+		 * of async reads queued on vdev queue. So, account for prefetch
+		 * issued for blocks pointed by indirect and limit max prefetch
+		 * in one go to zfs_traverse_indirect_prefetch_limit.
+		 */
+		prefetchtriggeridx = 0;
+		prefetchidx = 0;
+		prefetched = 0;
+		prefetchlimit = MIN(epb, zfs_traverse_indirect_prefetch_limit);
+		ASSERT(prefetchtriggeridx <= prefetchidx);
 		for (i = 0; i < epb; i++) {
-			SET_BOOKMARK(czb, zb->zb_objset, zb->zb_object,
-			    zb->zb_level - 1,
-			    zb->zb_blkid * epb + i);
-			traverse_prefetch_metadata(td,
-			    &((blkptr_t *)buf->b_data)[i], czb);
-		}
-
-		/* recursively visitbp() blocks below this */
-		for (i = 0; i < epb; i++) {
+			if (prefetchlimit && i == prefetchtriggeridx &&
+			    prefetchidx < epb) {
+				ASSERT(prefetchtriggeridx <= prefetchidx);
+				prefetched = 0;
```

Comment:
This could be moved inside for, considering below.

Reply:
Not sure what you mean with below here, but I mean `for (prefetched = 0; ...`.

Reply:
By Bad. would move it inside for.